k256/src/arithmetic/affine_op.rs

```diff
@@ -0,0 +1,423 @@
+
+    fn add(self, other: PowdrAffinePoint) -> PowdrAffinePoint {
+        if self.0.infinity != 0 {
+            return other;
+        }
+        if other.0.infinity != 0 {
+            return self;
+        }
+
+        if other.0.x == self.0.x {
+            if self.0.y == other.0.y {
+                return self.double();
+            } else {
+                //  x1 == x2 but y1 != y2 → vertical line → point at infinity
+                return PowdrAffinePoint(AffinePoint::IDENTITY);
+            }
+        }
+
+        // normalization is needed here to ensure the initial magnitude is 1.
+        let x1 = self.0.x.normalize_weak();
+        let y1 = self.0.y.normalize_weak();
+
+        let dx = other.0.x - x1;
+        let invert = dx.invert().unwrap();
+
+        let dy = other.0.y - y1;
+        let lambda = dy * invert;
+
+        let x3 = lambda.square() - x1 - other.0.x;
+        let y3 = lambda * (x1 + x3.negate(5)) - y1;
+
+        PowdrAffinePoint(AffinePoint {
+            x: x3.normalize_weak(),
+            y: y3.normalize_weak(),
+            infinity: 0,
+        })
+    }
+}
+
+impl Neg for PowdrAffinePoint {
+    type Output = PowdrAffinePoint;
+
+    fn neg(self) -> PowdrAffinePoint {
+        PowdrAffinePoint::neg(&self)
+    }
+}
+
+impl Mul<Scalar> for PowdrAffinePoint {
+    type Output = PowdrAffinePoint;
+
+    fn mul(self, other: Scalar) -> PowdrAffinePoint {
+        mul(&self, &other)
+    }
+}
+
+impl PowdrAffinePoint {
+    /// Ruduces the coordinates of the point to their canonical form.
+    pub fn normalize_coordinates(&self) -> Self {
+        PowdrAffinePoint(AffinePoint {
+            x: self.0.x.normalize(),
+            y: self.0.y.normalize(),
+            infinity: self.0.infinity,
+        })
+    }
+
+    /// Double the point.
+    pub fn double(self) -> PowdrAffinePoint {
+        if self.0.y.normalizes_to_zero().into() {
+            return PowdrAffinePoint(AffinePoint::IDENTITY);
+        }
+
+        let num = FieldElement::from(3u64) * self.0.x.square();
+        let denom = FieldElement::from(2u64) * self.0.y;
+        let lambda = num * denom.invert().unwrap();
+
+        let x3 = lambda.square() - FieldElement::from(2u64) * self.0.x;
+        let y3 = lambda * (self.0.x + x3.negate(3)) - self.0.y;
+
+        PowdrAffinePoint(AffinePoint {
+            x: x3.normalize_weak(),
+            y: y3.normalize_weak(),
+            infinity: 0,
+        })
+    }
+
+    fn neg(&self) -> Self {
+        PowdrAffinePoint(AffinePoint {
+            x: self.0.x,
+            y: self.0.y.negate(1).normalize(),
+            infinity: self.0.infinity,
+        })
+    }
+
+    fn conditional_select(a: &Self, b: &Self, choice: Choice) -> Self {
+        PowdrAffinePoint(AffinePoint {
+            x: FieldElement::conditional_select(&a.0.x, &b.0.x, choice),
+            y: FieldElement::conditional_select(&a.0.y, &b.0.y, choice),
+            infinity: a.0.infinity,
+        })
```

Comment:
```suggestion
        PowdrAffinePoint(AffinePoint::conditional_select(&a.0, &b.0, choice))
```
You could also implement the `ConditionallySelectable` trait for `PowdrAffinePoint`.

Reply:
Done